src/btree/bt_cursor.c

```diff
@@ -2222,6 +2222,135 @@ __wt_btcur_range_truncate(WT_TRUNCATE_INFO *trunc_info)
+        }
+    else if (collator == NULL) {
+        /*
+         * In some cases we expect we're comparing more than a few keys with matching prefixes, so
+         * it's faster to avoid the memory fetches by skipping over those prefixes. That's done by
+         * tracking the length of the prefix match for the lowest and highest keys we've seen
+         * previously.
+         *
+         * Normally we'd expect every parent page's skippable prefixes to be shorter than the
+         * prefixes we can skip in the child page, and so we'd skip increasingly longer prefixes as
+         * we walk down the tree (in other words, if we can skip N bytes on the parent, we can skip
+         * at least N bytes on the child). However, if the search threads cache this skippable
+         * prefix size as they move down the tree, and if the tree structure changes in parallel -
+         * for example page splits reducing the child pages key space or a keys destined for a
+         * now-deleted sibling page being inserted into the current page - the skippable prefix can
+         * be incorrect for the page. To protect against this we reset the skippable prefix length
+         * each time we move to a new page.
+         */
+        skiphigh = skiplow = 0;
+
+        for (; limit != 0; limit >>= 1) {
+            indx = base + (limit >> 1);
+            rip = page->pg_row + indx;
+            WT_ERR(__wt_row_leaf_key(session, page, rip, item, true));
+
+            match = WT_MIN(skiplow, skiphigh);
+            cmp = __wt_lex_compare_skip(session, srch_key, item, &match);
+            if (cmp > 0) {
+                skiplow = match;
+                base = indx + 1;
+                --limit;
+            } else if (cmp < 0)
+                skiphigh = match;
+            else
+                goto leaf_match;
+        }
+    } else {
+        WT_ERR_MSG(session, WT_ERROR, "Collator not supported!");
+    }
+
+    /*
+     * The best case is finding an exact match in the leaf page's WT_ROW array, probable for any
+     * read-mostly workload. Check that case and get out fast.
+     */
+    if (0) {
+leaf_match:
+        // TODO: handle skip list entries?
+        *indx_ptr = WT_ROW_SLOT(page, rip);
+        return (0);
+    }
+
+    /*
+     * We didn't find an exact match in the WT_ROW array.
+     *
+     * Base is the smallest index greater than key and may be the 0th index or the (last + 1) index.
+     * If base is the 0th index it means the key is before any key found on the page.
+     */
+    if (base == 0) {
+        *indx_ptr = 0;
+
+        F_SET(cbt, WT_CBT_SEARCH_SMALLEST);
+        ins_head = WT_ROW_INSERT_SMALLEST(page);
+    } else {
+        *indx_ptr = base - 1;
+
+        ins_head = WT_ROW_INSERT_SLOT(page, cbt->slot);
+    }
+
+    if (WT_SKIP_FIRST(ins_head) == NULL)
+        return (0);
+    
+    WT_ERR_MSG(session, WT_ERROR, "Found a leaf with pending insert entries. Did you forget to checkpoint?");
+    return (0);
```

Comment:
Not sure if the return value should be non-zero in this error case. But this probably isn't super important for the spike, so I'm fine to leave as is

Reply:
Oh wait, I think that if we hit the error, we go to the `err` label where we return non-zero.

Reply:
Exactly, so the return is never hit. I removed the line.